core/include/tpm2.h

```diff
@@ -0,0 +1,39 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2022, Linaro Limited
+ */
+
+#ifndef __TPM2_H__
+#define __TPM2_H__
+
+#include <stdint.h>
+#include <types_ext.h>
+#include <util.h>
+
+/* TPM2_ST Structure Tags */
+#define TPM2_ST_RSP_COMMAND	U(0x00C4)
+#define TPM2_ST_NULL		U(0X8000)
+#define TPM2_ST_NO_SESSIONS	U(0x8001)
+#define TPM2_ST_SESSIONS	U(0x8002)
+
+/* TPM2_SU Constants Shutdown and startup modes */
+#define TPM2_SU_CLEAR		U(0x0000)
+#define TPM2_SU_STATE		U(0x0001)
+
+/* Command Codes */
+#define	TPM2_CC_NV_WRITE	U(0x00000137)
+#define	TPM2_CC_SELFTEST	U(0x00000143)
+#define TPM2_CC_STARTUP		U(0x00000144)
+#define	TPM2_CC_NV_READ		U(0x0000014E)
+#define	TPM2_CC_GET_CAPABILITY  U(0x0000017A)
+#define	TPM2_CC_PCR_READ	U(0x0000017E)
+#define	TPM2_CC_PCR_EXTEND	U(0x00000182)
+
+/* Command Error Codes */
+#define TPM2_RC_SUCCESS		0x0
+
+int tpm2_startup(uint16_t mode);
+int tpm2_selftest(uint8_t full);
```

Comment:
Could use `enum tpm2_result` or `TEE_Result` error code.

Reply:
I am using int as want to distinguish between tpm interface errors and command related errors.. Have added a description for these commands...Let me know if that looks ok..

Reply:
Looks a bit strange. This adds a specific positive TPM2 error core, and negative... system error code.
I think this function should preferably return either a `enum tpm2_result` or `TEE_Result` to avoid confusion.
If `enum tpm2_result`: `TPM2_ERR_GENERIC` could match any drive rissue (out-of-mem etc..).
If `TEE_Result`: `TEE_ERROR_COMMUNICATION` could match any TPM2 comm failures.

I think `tpm2_startup()` should also return either a `TEE_Result` or a `enum tpm2_result`

Reply:
Changing it to tpm2_result.

I am also rewriting these functions to always pass a command and response buffer. These buffers can be on stack for startup and selftest as they are going to be small